File: analysis/ham_cycle_search.cpp
/**
 * ham_cycle_search.cpp
 * 
 * Direction 1: Hamiltonian cycle GA search in permutation space.
 * 
 * Encodes a solution as a permutation π of {0, ..., n-1} where row i
 * has points at columns (π[i], π[i+1]). This guarantees 2 per row/col.
 * 
 * Search: simulated annealing with targeted 2-opt swaps.
 * Compile: g++ -O3 -march=native -std=c++17 -static ham_cycle_search.cpp -o ham_cycle_search
 * Run: ./ham_cycle_search <n> [restarts] [iters]
 */

#include <cstdio>
#include <cstdint>
#include <cstdlib>
#include <cmath>
#include <algorithm>
#include <vector>
#include <random>
#include <chrono>
#include <cstring>
using namespace std;

int N;
mt19937 rng(chrono::steady_clock::now().time_since_epoch().count());

// Distance cache for each grid cell
vector<int> dist_cache;

void build_dist_cache() {
    dist_cache.resize(N * N);
    int cx2 = N - 1, cy2 = N - 1;
    for (int x = 0; x < N; x++)
        for (int y = 0; y < N; y++) {
            int dx = 2 * x - cx2;
            int dy = 2 * y - cy2;
            dist_cache[y * N + x] = dx * dx + dy * dy;
        }
}

inline int get_dist(int x, int y) {
    return dist_cache[y * N + x];
}

// Compute fitness for a permutation
// Returns (collinear_count, ring_overload, total = collinear*1000 + overload)
struct Fitness {
    int collinear;
    int overload;
    int total;
};

Fitness compute_fitness(const vector<int> &perm) {
    // Use forbid_accum to count collinear triples
    vector<uint64_t> forbid(N, 0);
    int collinear = 0;
    
    for (int r = 0; r < N; r++) {
        int c1 = perm[r];
        int c2 = perm[(r + 1) % N];
        
        if ((forbid[r] >> c1) & 1) collinear++;
        if ((forbid[r] >> c2) & 1) collinear++;
        
        // Update forbid for future rows
        for (int pr = 0; pr < r; pr++) {
            int pc1 = perm[pr];
            int pc2 = perm[(pr + 1) % N];
            
            int pairs[4][2] = {{pc1, c1}, {pc1, c2}, {pc2, c1}, {pc2, c2}};
            for (auto &p : pairs) {
                int pc = p[0], cc = p[1];
                int dr = r - pr;
                int dc = cc - pc;
                if (dr == 0) continue;
                for (int tr = r + 1; tr < N; tr++) {
                    int num = dc * (tr - pr);
                    if (num % dr == 0) {
                        int col = pc + num / dr;
                        if (col >= 0 && col < N)
                            forbid[tr] |= (1ULL << col);
                    }
                }
            }
        }
    }
    
    // Ring overload
    vector<int> ring_counts(2000, 0);  // d values are bounded by ~2*(N-1)²
    for (int r = 0; r < N; r++) {
        int c1 = perm[r];
        int c2 = perm[(r + 1) % N];
        ring_counts[get_dist(c1, r)]++;
        ring_counts[get_dist(c2, r)]++;
    }
    
    int overload = 0;
    for (int c : ring_counts)
        if (c > 2) overload += c - 2;
    
    return {collinear, overload, collinear * 1000 + overload};
}

// Find which rows are problematic
void find_problem_rows(const vector<int> &perm, vector<int> &problem_rows) {
    vector<uint64_t> forbid(N, 0);
    vector<int> ring_counts(2000, 0);
    vector<bool> has_collinear(N, false);
    vector<bool> has_overload(N, false);
    
    // Track ring counts
    for (int r = 0; r < N; r++) {
        int c1 = perm[r];
        int c2 = perm[(r + 1) % N];
        ring_counts[get_dist(c1, r)]++;
        ring_counts[get_dist(c2, r)]++;
    }
    
    // Find overloaded rows
    for (int r = 0; r < N; r++) {
        int c1 = perm[r];
        int c2 = perm[(r + 1) % N];
        if (ring_counts[get_dist(c1, r)] > 2) has_overload[r] = true;
        if (ring_counts[get_dist(c2, r)] > 2) has_overload[r] = true;
    }
    
    // Find collinear rows
    for (int r = 0; r < N; r++) {
        int c1 = perm[r];
        int c2 = perm[(r + 1) % N];
        
        if ((forbid[r] >> c1) & 1) has_collinear[r] = true;
        if ((forbid[r] >> c2) & 1) has_collinear[r] = true;
        
        for (int pr = 0; pr < r; pr++) {
            int pc1 = perm[pr];
            int pc2 = perm[(pr + 1) % N];
            int pairs[4][2] = {{pc1, c1}, {pc1, c2}, {pc2, c1}, {pc2, c2}};
            for (auto &p : pairs) {
                int pc = p[0], cc = p[1];
                int dr = r - pr;
                int dc = cc - pc;
                if (dr == 0) continue;
                for (int tr = r + 1; tr < N; tr++) {
                    int num = dc * (tr - pr);
                    if (num % dr == 0) {
                        int col = pc + num / dr;
                        if (col >= 0 && col < N)
                            forbid[tr] |= (1ULL << col);
                    }
                }
            }
        }
    }
    
    problem_rows.clear();
    for (int r = 0; r < N; r++)
        if (has_collinear[r] || has_overload[r])
            problem_rows.push_back(r);
}

// Simulated annealing search
vector<int> search(int restarts, int max_iters) {
    vector<int> best_perm;
    int best_total = INT_MAX;
    
    auto t0 = chrono::high_resolution_clock::now();
    
    for (int restart = 0; restart < restarts; restart++) {
        // Random initial permutation
        vector<int> perm(N);
        iota(perm.begin(), perm.end(), 0);
        shuffle(perm.begin(), perm.end(), rng);
        
        Fitness fit = compute_fitness(perm);
        int current = fit.total;
        
        if (current < best_total) {
            best_total = current;
            best_perm = perm;
        }
        
        if (current == 0) {
            auto t1 = chrono::high_resolution_clock::now();
            double sec = chrono::duration<double>(t1 - t0).count();
            printf("✅ Found at restart %d (%.1fs)\n", restart + 1, sec);
            return perm;
        }
        
        double temp_start = 10.0;
        double temp_end = 0.01;
        
        for (int iter = 0; iter < max_iters; iter++) {
            double T = temp_start * pow(temp_end / temp_start, (double)iter / max_iters);
            
            // Find problem rows
            vector<int> problem_rows;
            find_problem_rows(perm, problem_rows);
            
            // Targeted swap
            int i, j;
            if (problem_rows.empty()) {
                i = rng() % N;
                j = rng() % N;
                while (j == i) j = rng() % N;
            } else {
                i = problem_rows[rng() % problem_rows.size()];
                j = rng() % N;
                while (j == i) j = rng() % N;
            }
            
            vector<int> new_perm = perm;
            swap(new_perm[i], new_perm[j]);
            
            Fitness new_fit = compute_fitness(new_perm);
            int delta = new_fit.total - current;
            
            if (delta < 0 || (double)rng() / rng.max() < exp(-delta / max(T, 0.001))) {
                perm = new_perm;
                current = new_fit.total;
                
                if (current < best_total) {
                    best_total = current;
                    best_perm = perm;
                    
                    if (current == 0) {
                        auto t1 = chrono::high_resolution_clock::now();
                        double sec = chrono::duration<double>(t1 - t0).count();
                        printf("✅ Found at restart %d, iter %d (%.1fs)\n", restart + 1, iter, sec);
                        return perm;
                    }
                }
            }
        }
        
        if ((restart + 1) % 5 == 0) {
            auto t1 = chrono::high_resolution_clock::now();
            double sec = chrono::duration<double>(t1 - t0).count();
            printf("R%d/%d: best=%d [%.1fs]\n", restart + 1, restarts, best_total, sec);
        }
    }
    
    auto t1 = chrono::high_resolution_clock::now();
    double sec = chrono::duration<double>(t1 - t0).count();
    printf("❌ Not found in %d restarts [%.1fs]. Best fitness: %d\n", restarts, sec, best_total);
    
    return best_perm;
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <n> [restarts] [iters]\n", argv[0]);
        return 1;
    }
    
    N = atoi(argv[1]);
    int restarts = (argc > 2) ? atoi(argv[2]) : 200;
    int iters = (argc > 3) ? atoi(argv[3]) : 20000;
    
    build_dist_cache();
    
    // Compute permutation count
    long long perm_count = 1;
    for (int i = 2; i <= N; i++) perm_count *= i;
    
    printf("Hamiltonian Cycle Search: n=%d\n", N);
    printf("  Restarts: %d, Iters/restart: %d\n", restarts, iters);
    printf("  Perm space: %lld possibilities\n", perm_count);
    printf("\n");
    
    vector<int> result = search(restarts, iters);
    
    if (!result.empty()) {
        Fitness fit = compute_fitness(result);
        printf("\nResult: fitness=%d\n", fit.total);
        printf("  Collinear: %d\n", fit.collinear);
        printf("  Ring overload: %d\n", fit.overload);
        
        if (fit.total == 0) {
            printf("  Permutation: [");
            for (int i = 0; i < N; i++)
                printf("%d%c", result[i], i < N-1 ? ',' : ']');
            printf("\n  Solution:\n");
            for (int r = 0; r < N; r++) {
                int c1 = result[r];
                int c2 = result[(r + 1) % N];
                printf("    Row %2d: (%d,%d) d=%d, (%d,%d) d=%d\n",
                       r, c1, r, get_dist(c1, r), c2, r, get_dist(c2, r));
            }
        }
    }
    
    return 0;
}

File: analysis/ham_cycle_search.py
#!/usr/bin/env python3
"""
方向一：哈密顿循环置换空间搜索 (Hamiltonian Cycle GA Search)

把解编码为一个置换 π[0..n-1]，行 i 放置 (π[i], i) 和 (π[i+1], i)。
这个编码天然满足：
- 每行 2 点 ✓
- 每列 2 点 ✓  

搜索目标：找到一个置换使得：
- 无三点共线 (collinearity = 0)
- 每个距离环 ≤ 2 点 (missing-center)

使用 2-opt 局部搜索 + 随机重启。

Usage:
    python ham_cycle_search.py <n> [max_iterations] [restarts]
"""

import sys
import random
from collections import defaultdict, Counter
from itertools import combinations
import math
import time


class HamiltonianSolution:
    """A No-Three-In-Line solution encoded as a column permutation."""
    
    def __init__(self, n, perm=None):
        self.n = n
        self.cx2 = self.cy2 = n - 1
        self.perm = perm if perm else self._random_perm()
        self._cached_fitness = None
    
    def _random_perm(self):
        """Generate a random permutation of 0..n-1."""
        p = list(range(self.n))
        random.shuffle(p)
        return p
    
    def get_points(self):
        """Return list of (x, y) = (col, row) for this solution."""
        pts = []
        for i in range(self.n):
            c1 = self.perm[i]
            c2 = self.perm[(i + 1) % self.n]
            pts.append((c1, i))
            pts.append((c2, i))
        return pts
    
    def get_dist(self, x, y):
        """Squared distance from grid center."""
        dx = 2 * x - self.cx2
        dy = 2 * y - self.cy2
        return dx * dx + dy * dy
    
    def fitness(self):
        """Compute fitness: lower is better.
        0 = perfect solution (no collinearity, missing-center).
        """
        if self._cached_fitness is not None:
            return self._cached_fitness
        
        pts = self.get_points()
        k = len(pts)
        
        # Penalty 1: Collinearity
        collinear_count = 0
        for i in range(k):
            x1, y1 = pts[i]
            for j in range(i + 1, k):
                x2, y2 = pts[j]
                for m in range(j + 1, k):
                    x3, y3 = pts[m]
                    # Cross product = 0 means collinear
                    area2 = (x2 - x1) * (y3 - y1) - (x3 - x1) * (y2 - y1)
                    if area2 == 0:
                        collinear_count += 1
        
        # Penalty 2: Ring overload
        ring_counts = Counter()
        for x, y in pts:
            d = self.get_dist(x, y)
            ring_counts[d] += 1
        
        ring_overload = sum(max(0, c - 2) for c in ring_counts.values())
        
        # Heavier weight on collinearity (must be 0 for valid solution)
        fitness = collinear_count * 1000 + ring_overload
        
        self._cached_fitness = fitness
        return fitness
    
    def is_valid(self):
        """Check if this is a valid missing-center solution."""
        return self.fitness() == 0
    
    def swap(self, i, j):
        """Swap positions i and j in the permutation (2-opt)."""
        new_perm = list(self.perm)
        new_perm[i], new_perm[j] = new_perm[j], new_perm[i]
        return HamiltonianSolution(self.n, new_perm)
    
    def invert(self, i, j):
        """Reverse segment [i, j] in the permutation."""
        new_perm = list(self.perm)
        new_perm[i:j+1] = reversed(new_perm[i:j+1])
        return HamiltonianSolution(self.n, new_perm)
    
    def neighbor(self):
        """Generate a random neighbor by 2-opt swap."""
        i, j = random.sample(range(self.n), 2)
        return self.swap(i, j)
    
    def __str__(self):
        pts = self.get_points()
        f = self.fitness()
        is_valid = self.is_valid()
        return f"perm={self.perm}, fitness={f}, valid={'✅' if is_valid else '❌'}"


def local_search(initial, max_iterations=10000, patience=500):
    """Hill climbing with random neighbor and restart."""
    current = initial
    best = initial
    best_fitness = initial.fitness()
    
    iterations_without_improvement = 0
    
    for iteration in range(max_iterations):
        neighbor = current.neighbor()
        nf = neighbor.fitness()
        
        if nf < best_fitness:
            best = neighbor
            best_fitness = nf
            iterations_without_improvement = 0
            
            if nf == 0:
                return best, iteration, True
        
        if nf <= current.fitness():
            current = neighbor
        
        iterations_without_improvement += 1
        if iterations_without_improvement >= patience:
            # Restart from a new random solution
            current = HamiltonianSolution(current.n)
            iterations_without_improvement = 0
    
    return best, max_iterations, best_fitness == 0


def search(n, restarts=100, max_iter=10000, patience=500):
    """Main search: multiple restarts with local search."""
    print(f"Searching n={n}: {restarts} restarts, {max_iter} iter each")
    print(f"  Permutation space: {n}! possibilities")
    print()
    
    best_ever = None
    best_fitness_ever = float('inf')
    total_attempts = 0
    start_time = time.time()
    
    for restart in range(restarts):
        initial = HamiltonianSolution(n)
        best_sol, iters, found = local_search(initial, max_iter, patience)
        total_attempts += iters
        
        bf = best_sol.fitness()
        if bf < best_fitness_ever:
            best_fitness_ever = bf
            best_ever = best_sol
            print(f"  Restart {restart+1}: fitness improved to {bf} (iter={iters})")
            
            if found:
                elapsed = time.time() - start_time
                print(f"\n✅ FOUND VALID SOLUTION! ({elapsed:.1f}s)")
                print(f"  Permutation: {best_ever.perm}")
                print(f"  Points: {sorted(best_ever.get_points())}")
                return best_ever
        
        if (restart + 1) % 10 == 0:
            elapsed = time.time() - start_time
            print(f"  [{restart+1}/{restarts}] best={best_fitness_ever}, {elapsed:.0f}s")
    
    elapsed = time.time() - start_time
    print(f"\nNo valid solution found in {restarts} restarts ({elapsed:.1f}s)")
    print(f"Best fitness: {best_fitness_ever}")
    
    if best_ever:
        print(f"Best permutation: {best_ever.perm}")
        # Analyze what went wrong with the best solution
        pts = best_ever.get_points()
        ring_counts = Counter()
        for x, y in pts:
            d = best_ever.get_dist(x, y)
            ring_counts[d] += 1
        print(f"Ring counts: {dict(sorted(ring_counts.items()))}")
    
    return best_ever


def exhaustive_perm_search(n):
    """Try ALL permutations (only for small n like n ≤ 6)."""
    import itertools
    
    print(f"Exhaustive search n={n}...")
    found = []
    
    for idx, perm in enumerate(itertools.permutations(range(n))):
        sol = HamiltonianSolution(n, list(perm))
        if sol.is_valid():
            found.append(perm)
            print(f"  #{len(found)}: {perm}")
    
    print(f"Found {len(found)} valid solutions")
    return found


if __name__ == '__main__':
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 12
    restarts = int(sys.argv[2]) if len(sys.argv) > 2 else 200
    max_iter = int(sys.argv[3]) if len(sys.argv) > 3 else 5000
    
    if n <= 6:
        # For small n, exhaustive search is feasible
        exhaustive_perm_search(n)
    else:
        sol = search(n, restarts=restarts, max_iter=max_iter)
        if sol and sol.is_valid():
            # Verify
            pts = sol.get_points()
            print(f"\nVerification:")
            ok, msg = True, "OK"
            for i in range(len(pts)):
                for j in range(i+1, len(pts)):
                    for k in range(j+1, len(pts)):
                        x1,y1 = pts[i]; x2,y2 = pts[j]; x3,y3 = pts[k]
                        if (x2-x1)*(y3-y1) - (x3-x1)*(y2-y1) == 0:
                            ok = False
                            msg = f"Collinear: {pts[i]}, {pts[j]}, {pts[k]}"
                            break
                    if not ok: break
                if not ok: break
            print(f"  {'✅' if ok else '❌'} {msg}")
